File: risc0/zkp/core/poly.cpp
// Copyright 2022 Risc0, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "risc0/zkp/core/poly.h"

#include "risc0/core/log.h"

#include <vector>

namespace risc0 {

Fp4 polyEval(const Fp4* coeffs, size_t size, Fp4 val) {
  Fp4 tot;
  Fp4 mul(1, 0, 0, 0);
  for (size_t i = 0; i < size; i++) {
    tot += mul * coeffs[i];
    mul *= val;
  }
  return tot;
}

Fp4 polyDivide(Fp4* p, size_t size, Fp4 z) {
  Fp4 cur;
  for (size_t i = size; i-- > 0;) {
    Fp4 next = z * cur + p[i];
    p[i] = cur;
    cur = next;
  }
  return cur;
}

void polyInterpolate(Fp4* out, const Fp4* x, const Fp4* fx, size_t size) {
  // Special case the very easy ones
  if (size == 1) {
    out[0] = fx[0];
    return;
  }
  if (size == 2) {
    out[1] = (fx[1] - fx[0]) * inv(x[1] - x[0]);
    out[0] = fx[0] - out[1] * x[0];
    return;
  }
  // Compute ft = product of (x - x_i) for all i
  std::vector<Fp4> ft(size + 1);
  ft[0] = Fp4(1, 0, 0, 0);
  for (size_t i = 0; i < size; i++) {
    for (size_t j = i + 1; j-- > 0;) {
      ft[j + 1] += ft[j];
      ft[j] *= -x[i];
    }
  }
  // Clear output
  for (size_t i = 0; i < size; i++) {
    out[i] = Fp4();
  }
  for (size_t i = 0; i < size; i++) {
    // Compute fr = ft / (x - x_i)
    std::vector<Fp4> fr = ft;
    polyDivide(fr.data(), fr.size(), x[i]);
    // Evaluate at x[i]
    Fp4 fr_xi = polyEval(fr.data(), fr.size(), x[i]);
    // Compute multiplier (fx[i] / fr_xi)
    Fp4 mul = fx[i] * inv(fr_xi);
    // Multiply into output
    for (size_t j = 0; j < size; j++) {
      out[j] += mul * fr[j];
    }
  }
}

} // namespace risc0

File: risc0/zkp/src/core/poly.rs
// Copyright 2026 RISC Zero, Inc.
//
// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
// http://opensource.org/licenses/MIT>, at your option. This file may not be
// copied, modified, or distributed except according to those terms.
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

//! Polynomial utilities (currently only those used in polynomial evaluation).

use alloc::vec;

use risc0_core::field::ExtElem;

/// Evaluate a polynomial whose coefficients are in the extension field at a
/// point.
pub fn poly_eval<E: ExtElem>(coeffs: &[E], x: E) -> E {
    let mut mul = E::ONE;
    let mut tot = E::ZERO;
    for coeff in coeffs {
        tot += *coeff * mul;
        mul *= x;
    }
    tot
}

/// General purpose polynomial interpolation.
///
/// Given the goal value f(x) at a set of evaluation points x, compute
/// coefficients.
pub fn poly_interpolate<E: ExtElem>(out: &mut [E], x: &[E], fx: &[E], size: usize) {
    // Special case the very easy ones
    if size == 1 {
        out[0] = fx[0];
        return;
    }
    if size == 2 {
        out[1] = (fx[1] - fx[0]) * (x[1] - x[0]).inv();
        out[0] = fx[0] - out[1] * x[0];
        return;
    }
    // Compute ft = product of (x - x_i) for all i
    let mut ft = vec![E::ZERO; size + 1];
    ft[0] = E::ONE;
    for (i, x) in x.iter().enumerate().take(size) {
        for j in (0..i + 1).rev() {
            let value = ft[j];
            ft[j + 1] += value;
            ft[j] *= -*x;
        }
    }
    // Clear output
    for x in &mut *out {
        *x = E::ZERO;
    }
    for i in 0..size {
        // Compute fr = ft / (x - x_i)
        let mut fr = ft.clone();
        poly_divide(&mut fr, x[i]);
        // Evaluate at x[i]
        let fr_xi = poly_eval(&fr, x[i]);
        // Compute multiplier (fx[i] / fr_xi)
        let mul = fx[i] * fr_xi.inv();
        // Multiply into output
        for j in 0..size {
            out[j] += mul * fr[j];
        }
    }
}

/// In-place polynomial division.
///
/// Take the coefficients in P, and divide by (X - z) for some z, return the
/// remainder.
pub fn poly_divide<E: ExtElem>(p: &mut [E], z: E) -> E {
    let mut cur = E::ZERO;
    for i in (0..p.len()).rev() {
        let next = z * cur + p[i];
        p[i] = cur;
        cur = next;
    }
    cur
}
